File: src/layout/widget/feature/Feature.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package layout.widget.feature;

import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.TYPE;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Provides metadata information about an interface as a feature of an object
 * implementing it. The object can then provide better information about itself
 *  (its features) by providing information about the interfaces it implements.
 * <p>
 * Simply, if implementing an interface tells compiler what an object can do,
 * annotating the interface as a Feature tells the same information to the
 * user of the application.
 *
 * @author Plutonium_
 */
@Retention(RUNTIME)
@Target(TYPE)
public @interface Feature {
    /** Identifies the feature. Not necessarily unique. */
    String name() default "";
    /** Description of the feature.. */
    String description() default "";
    /**
     * Identifies the feature exactly. Must be unique and must match the class of the annotated
     * interface.
     */
    Class type() default Void.class;
}

File: src/player/main/sp/it/pl/layout/widget/feature/Feature.kt
package sp.it.pl.layout.widget.feature

import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.CLASS
import kotlin.reflect.KClass

/**
 * Provides metadata information about an interface as a feature of an object implementing it.
 *
 * While implementing an interface tells compiler what an object can do, annotating the interface as a Feature then
 * tells this information to the user.
 *
 * Feature is not and can not be expressed by the type system (be part of the interface's signature), because each
 * feature carries its own information and multi-inheritance would not allow this.
 */
@Retention(RUNTIME)
@Target(CLASS)
annotation class Feature(
    /** Identifies the feature. Not necessarily unique. */
    val name: String = "",
    /** Description of the feature. */
    val description: String = "",
    /** Identifies the feature exactly. Must be unique and must be the class of the annotated interface. */
    val type: KClass<*>
)